feat: Disable Sharing globally

marination opened the pull request:
Docs: https://docs.erpnext.com/docs/v14/user/manual/en/setting-up/settings/system-settings/edit-wiki?wiki_page_patch=57c20f0965

### Issue:
- The Document Share feature is a way to essentially go around the set permissions and give access to a user
- In many cases this is not desirable as the permissions are thought through well and a confidential document must not be viewed by anyone other than the permitted users.
- A case was encountered where a user shared a document with another user who was not supposed to see the document. This could happen accidentally or with sinister motives.
- Users can "accidentally" assign a document to a non-permitted user and share access again in the process.

### Enable or Disable document sharing system wide
1. Checkbox in System Settings which is unchecked by default to maintain old behaviour
    <img width="1329" alt="Screenshot 2023-03-14 at 8 11 34 PM" src="https://user-images.githubusercontent.com/25857446/225041119-a1f913e2-6554-4724-8e58-edc1c2e3eb7c.png">

2. If enabled, system wide, no user will have share access to any document
3. The share button will be read only, no action (no share access). Although users can see who the doc is shared with (consistent with old behaviour)
     ![2023-03-14 20 26 04](https://user-images.githubusercontent.com/25857446/225041732-99ea1756-2445-4d9d-a548-d87b201bfaad.gif)

4. On invoking the share API **without** `ignore_share_permissions` (form the browser), a perm error is thrown
    ![2023-03-14 20 33 34](https://user-images.githubusercontent.com/25857446/225044263-75334d2d-0988-4a82-bf64-da69dd4d18c7.gif)
    
5. If any share API is invoked **with** `ignore_share_permissions` it will go through (consistent with old behaviour). [Creating a user, doc must be shared with the user themselves]

   ![2023-03-15 12 04 25](https://user-images.githubusercontent.com/25857446/225226052-94829f0c-ac23-49ed-b0dc-1abc2d945424.gif)

6. `assign_to` is handled as a special case, where assigning a doc to a user **will block** the assignment if the user does not have access already. The assignee is expected to have access to the document otherwise assignments can be misused
    <img width="1299" alt="Screenshot 2023-03-28 at 5 57 31 PM" src="https://user-images.githubusercontent.com/25857446/228237899-d3327240-d229-43a4-843a-8362c1004388.png">

marination committed:
feat: Disable Sharing globally

- Checkbox in System Settings
- If disabled, avoid share UI render
- Share APIs return None (non-obstructing) if share APIs are invoked

barredterra commented on frappe/core/doctype/system_settings/system_settings.json:
```diff
@@ -528,12 +529,19 @@
+  {
+   "default": "0",
+   "description": "Strictly follow configured permissions",
+   "fieldname": "disable_sharing",
```
```suggestion
   "fieldname": "disable_access_sharing",
```

or

```suggestion
   "fieldname": "disable_permission_sharing",
```

to avoid future conflict with other sharing features?

gavindsouza replied:
I think `disable_document_sharing` would be more explicit (/apt) in this case.

barredterra replied:
Are we sharing the document or the permission to access it? 😁

marination replied:
Since it's referred in the docs as document sharing in multiple places, that seems like the best option. Users are not familiar with this feature as access/permission sharing. `DocShare` also gives that impression 😄
<img width="700" alt="Screenshot 2023-03-14 at 3 42 38 PM" src="https://user-images.githubusercontent.com/25857446/224968810-c96d0019-6546-4925-9545-23c761dc77ca.png">

<img width="500" alt="Screenshot 2023-03-14 at 3 44 54 PM" src="https://user-images.githubusercontent.com/25857446/224969188-15227621-7029-4f88-a502-1627dc7af3fd.png">

barredterra replied:
Well, well... I surrender 😄

marination committed:
feat: Settings checkbox must toggle share permission globally

- Treat feature like a perm toggler. Essentially noone is allowed to explicity share anything
- Implicit sharing via `ignore_share_permissions` is allowed. Devs can decide where sharing should happen under the hood
- UI is made read only and not hidden. Users must see who doc is already shared with
- Make sure perm APIs used by share feature return false if sharing is disabled
- Rename checkbox to `Disable Document Sharing`

marination committed:
Merge branch 'develop' into disable-sharing-globally

marination committed:
test: (server side) Impact of disabling sharing on APIs

- Also, fix missed system setting rename in `assign_to`

marination committed:
Merge branch 'develop' into disable-sharing-globally

marination committed:
Merge branch 'develop' into disable-sharing-globally

ankush commented on frappe/desk/form/assign_to.py:
```diff
@@ -93,8 +93,9 @@ def add(args=None):
-			if not frappe.has_permission(doc=doc, user=assign_to):
+			# if sharing is not disabled & assignee does not have permissions, share
+			sharing_disabled = frappe.get_system_settings("disable_document_sharing")
+			if not sharing_disabled and not frappe.has_permission(doc=doc, user=assign_to):
```
Let's at least tell the assigner if the assignee doesn't have permission? Improves UX. A msgprint or alert should be fine. 

@marination

marination replied:
@ankush done
<img width="1290" alt="Screenshot 2023-03-24 at 3 05 13 PM" src="https://user-images.githubusercontent.com/25857446/227483005-04a9aca1-9f22-4774-b56f-cd96baf31dac.png">

barredterra replied:
Does it make sense to allow assignment if the user cannot access it? Or should we `throw` instead of `msgprint`? I could imagine that the assigned user might still get confused when they get a todo without access. Just a thought, no strong preference here.

marination replied:
@barredterra I mean so far we made sure the user can see the document after being assigned, I guess `throw` would be consistent with that behaviour and would enforce that the permissions be given before assignment.
But on the other hand also blocking assignment because sharing is disabled feels a bit much. This depends on how often a user without access is assigned 😅
@ankush what do you think ? I'm conflicted

ankush replied:
@marination throw is fine too. :+1: Anything that avoids future confusion.

ankush reviewed (changes requested):
just one minor UX fix. Rest looks fine.

marination commented:
@ankush @shariquerik mind taking a look? Anything that majorly needs to change ?

marination committed:
fix: Inform assigner if assignee lacks perms and sharing is disabled

- misc: readable conditions

marination committed:
Merge branch 'develop' into disable-sharing-globally

ankush committed:
fix: throw instead of msgprint

ankush committed:
Merge branch 'develop' into disable-sharing-globally

marination committed:
fix: Typo and appropriate message for `throw`

barredterra commented on frappe/desk/form/assign_to.py:
```diff
@@ -93,10 +93,17 @@ def add(args=None):
+					msg += "<br>" + _(
+						"As document sharing is disabled, please give them the required permissions before assigning."
+					)
+					frappe.throw(msg, title="Missing Permission")  # nosemgrep
```
```suggestion
					frappe.throw(msg, title=_("Missing Permission"))
```

barredterra commented:
@ankush don't you want this in v13 at all or can we do a manual backport?

ankush commented:
Way too many conflicts, better to port manually and confirm if it works.

iMoshi commented:
Awesome! 

Would it not make more sense to actually hide the Share div from the sidebar when this feature is enabled though? :)

marination commented:
@iMoshi in the case where sharing is disabled after a document is already shared with some users, we would probably still want to see who all can view this document. Seemed consistent with the behaviour when a user has no share access.
That's what we were kinda going for - switch off everyones share permission.
